stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+use blockstack_lib::chainstate::nakamoto::NakamotoBlock;
+use blockstack_lib::chainstate::stacks::TenureChangePayload;
+use blockstack_lib::net::api::getsortition::SortitionInfo;
+use slog::{slog_info, slog_warn};
+use stacks_common::types::chainstate::{ConsensusHash, StacksPublicKey};
+use stacks_common::util::hash::Hash160;
+use stacks_common::{info, warn};
+
+use crate::client::{ClientError, StacksClient};
+use crate::signerdb::SignerDb;
+
+/// Captures this signer's current view of a sortition's miner.
+#[derive(PartialEq, Eq)]
+pub enum SortitionMinerStatus {
+    /// The signer thinks this sortition's miner is invalid, and hasn't signed any blocks for them.
+    InvalidatedBeforeFirstBlock,
+    /// The signer thinks this sortition's miner is invalid, but already signed one or more blocks for them.
+    InvalidatedAfterFirstBlock,
+    /// The signer thinks this sortition's miner is valid
+    Valid,
+}
+
+/// Captures the Stacks sortition related state for
+///  a successful sortition
+pub struct SortitionState {
+    /// The miner's pub key hash
+    pub miner_pkh: Hash160,
+    /// If known already, the public key which hashes to `miner_pkh`
+    pub miner_pubkey: Option<StacksPublicKey>,
+    /// the last burn block in this fork which had a sortition
+    pub prior_sortition: ConsensusHash,
+    /// the committed to parent tenure ID
+    pub parent_tenure_id: ConsensusHash,
+    /// this sortition's consensus hash
+    pub consensus_hash: ConsensusHash,
+    /// did the miner in this sortition do something
+    ///  to become invalidated as a miner?
+    pub invalidated: SortitionMinerStatus,
+}
+
+/// The signer's current view of the stacks chain's sortition
+///  state
+pub struct SortitionsView {
+    /// the prior successful sortition (this corresponds to the "prior" miner slot)
```

Comment:
(Note to self; may delete later)

As with `SortitionState`, it's going to be of paramount importance that `last_sortition`, `cur_sortition`, and `latest_consensus_hash`, if set, correspond to sortitions that all are ancestors of the Stacks node's canonical burnchain tip.  A Bitcoin reorg may invalidate this correspondence, and the signer will need to handle this case somehow (such as by treating any blocks and tenures -- even ones that are already signed -- as invalid once their corresponding `SortitionState`s or `SortitionsView`s get invalidated).

Reply:
All of this information is fetched from the stacks node’s RPC endpoint, which loads from the current canonical view. The view is refreshed in its entirety whenever a refresh is called, so it shouldn’t be possible for this struct to be split brained (unless the sortition DB was itself incoherent).

Reply:
Yup, I see that now.  Maybe add a docstring to indicate that this is refreshed via a node RPC whenever any important decisions regarding a block's validity get made?

Reply:
Docstring added in fdcaea8b91972ad721d9d9e0579ac89fb531ddc0